Specify null safety: Sections 'Variables' and 'Local Variable Declarations'

eernstg opened the pull request:
Leaf and Lasse, I believe you might discuss this PR and decide on whether one of you would give it a review. For now I'm just putting it on github such that reviewing can start whenever it fits (I guess it makes sense to wait for #2023).

This PR updates the sections 'Variables' and 'Local Variable Declarations' of the language specification to fit the language with null safety.

It is part of a larger PR, but provided in smaller pieces in order to make the reviewing task manageable. Because of this, the updated text contains a handful of undefined `\ref{...}` references. They refer to newly added sections which are included in https://github.com/dart-lang/language/pull/2023.

Apart from that, this PR changes `dart.sty` to a new version of the file which will work with all the null safety updates (so we don't have to update `dart.sty` again and again during this process). However, this causes a couple of changes outside the above mentioned sections: A couple of commands need to get an extra argument (it's an `{r}`, specifying the metasymbol which stands for either the empty string or `required`, which is now needed in some rules about functions and function types). Similarly, this PR also changes `\Gamma` to `\Delta` to denote type variable environments, because that's a more commonly used notation (and the new `dart.sty` has it in some commands already).

eernstg commented:
@lrhn, would you be able to take a look at this PR? I'm removing @leafpetersen as a reviewer, because it's obviously a good idea to broaden this effort a bit.

eernstg commented:
@lrhn, when you're back from vacation, could you take a look at this one?

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -17097,41 +17325,108 @@ \subsection{Local Variable Declaration}
+Assume that $D$ is a local variable declaration with the modifier \LATE{}
+that declares a variable $v$,
+which has an initializing expression $e$.
+It is a \Error{compile-time error} if $e$ contains an \AWAIT{} expression $a$
```
(We should not have allowed late local variables with initializers. 'Nuff said!)

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -17097,41 +17325,108 @@ \subsection{Local Variable Declaration}
+\commentary{%
+In other words,
+the initializing expression cannot await an expression directly,
+any await expressions must be nested inside some other function,
```
nested -> syntactically nested

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -17097,41 +17325,108 @@ \subsection{Local Variable Declaration}
+\begin{itemize}
+\item
+  If the static type of $e$ is \code{Null} then
+  the declared type of $v$ is \DYNAMIC.
```
I'm not sure `const myNull = null;` is valuable enough to justify spec updates. ;-)

But what would be the motivation for `late final x = expressionOfTypeNull;` to have static type `Null`?

For the change from `dynamic` to `Object?` we would need a breaking change request, it wouldn't be part of the specification of Dart 2.13.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -17097,41 +17325,108 @@ \subsection{Local Variable Declaration}
+  the declared type of $v$ is \code{X}.
+  \commentary{%
+    In this case $v$ is immediately promoted
+    (\ref{typePromotion}).%
```
Done.

eernstg reviewed (commented):
Thanks, and PTAL!

eernstg commented:
@lrhn, friendly ping.

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -17671,10 +18063,10 @@ \subsubsection{For-in}
 It follows that it is a compile-time error
-if $D$ is empty and \id{} is a final variable;
-and it is a dynamic error if $e$ has a top type,
+if $D$ is empty and \id{} is an immutable variable;
```
What if `id` is late and immutable, and hasn't been assigned yet, and the iterable has length at most one? Or just `final` and unassigned?

eernstg replied:
See the next comment.

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -17698,7 +18090,7 @@ \subsubsection{Asynchronous For-in}
 It is a dynamic type error if $o$ is not an instance of
 a class that implements \code{Stream}.
-It is a compile-time error if $D$ is empty and \id{} is a final variable.
+It is a compile-time error if $D$ is empty and \id{} is an immutable variable.
```
... is an immutable and non-`late` variable that has potentially been assigned, or if it is immutable and `late` and definitely assigned.

Should we have an abstraction?
> A local variable *may be assigned* if it is not immutable,
> or if it's immutable and definitely unassigned,
> or it's immutable and `late` and not definitely assigned.

(Variables with initializer expressions are always definitely assigned, even if they are late and not initialized yet.)

Just so we don't have to remember every case every time.

The following are both valid and run today:
```dart
  final int x;  
  for (x in [1]) {print(x);}
```
```dart
  late final int x;  
  for (x in [1]) {}
  print(x);
```

eernstg replied:
I get an error in DartPad for this:
```dart
void main() {
  final int x;
  for (x in [1]) {
    print(x);
  }
}
```
That would make sense, too, because there is no analysis which will establish that the for-statement iterates exactly one time. Also, it matches the treatment of the desugaring as a while-statement, where we are also not attempting to prove that it will execute the body at most once. I get the compile-time error for this one as well:

```dart
void main() {
  late final int x;
  for (x in [1]) {}
  print(x);
}
```

It looks like it is consistent with the implemented static analysis to say that it is an error when there is no declaration (so we're running a for-statement with an existing variable as the iteration variable), and that variable is immutable, late or non-late. The error messages could be improved, but we do flag these situations as errors.

So how did you run those two examples without encountering any errors?

lrhn replied:
The
```
line 8 * The late final local variable is already assigned.
```
looks spurious. At the assignment to `x2` in the loop, it's at most *possibly assigned*, and since it's `late`, that's OK.

That is, there is a possible execution sequence where the `late` variable doesn't throw when accessed, so it would be wrong to consider the program incorrect. (Non-late variable uses must be valid on *all* execution paths, late variable uses must be possibly valid on at least *one*.)

eernstg replied:
The error message for line 8 is of course less than ideal: It does not include the name of the variable.

I don't think we can use ad-hoc reasoning and conclude that a given late variable is possibly assigned, etc, because the actual analysis will do something specific (for loops, and for any other construct), and the resulting data is used to determine where/why to emit compile-time errors. We just have to live with the outcome, even when it is "easy to see" that some other result would have been more precise and still sound.

For example, the assignment to `x` in `for (x in ...)` is presumably approximated to run at least twice (even though it might just run once or not at all). This implies that we have approximated and then concluded that there is an assignment to `x` at a time where it is definitely assigned. That is indeed true for any loop execution that iterates more than one time.

However, I do think we should take this opportunity to double check that the implementations are consistent with the specification.

So we _should_ get an error for an `x` which is immutable, both when it is non-late and when it is late, and then both when it is definitely unassigned and when it is possibly assigned. We do get that:

```dart
void main() async {
  {
    // Immutable, non-late.
    final int x;
    await for (x in Stream.fromIterable([1])) {} // Error: can only set once.
    print(x); // Error: potentially unassigned: OK.
  }
  {
    // Immutable, late, definitely unassigned at `for`.
    late final int x;
    await for (x in Stream.fromIterable([1])) {} // Error: is already assigned.
    print(x);
  }
  {
    // Immutable, late, possibly assigned at `for`.
    late final int x;
    if (2 > 1) x = 1;
    await for (x in Stream.fromIterable([1])) {} // Error: is already assigned.
    print(x);
  }
}
```

I think we should just keep the spec as it is, given that this is also the implemented behavior.

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
 \LMHash{}%
-A \IndexCustom{final variable}{variable!final}
+An \IndexCustom{immutable variable}{variable!immutable}
 is a variable whose binding is fixed upon initialization;
```
Have we defined "initialization".

We use "bound to a value" below for most cases.
Does it cover both local and non-local variables?

eernstg commented on specification/dart.sty:
```diff
@@ -189,6 +205,22 @@
   {#1}_{#3}\ {#2}_{#3},\,\ldots,\ {#1}_{#4}\ {#2}_{#4}}}
 
+% Used to specify named arguments.
+% Parameters: Parameter name, argument name, index at start, index at end.
```
Done.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1276,36 +1284,53 @@ \section{Variables}
-It is a compile-time error for the declaration of
+It is a \Error{compile-time error} for the declaration of
 a variable which is not an instance variable
 to include the modifier \COVARIANT.
```
Good catch! Added commentary to make this distinction.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1276,36 +1284,53 @@ \section{Variables}
-\syntax{<metadata> <finalConstVarOrType>},
-we say that $v$ is the \Index{declaring occurrence} of the identifier.
+\syntax{<metadata> <finalConstVarOrType>} and \id{} is an identifier,
+we say that \id{} is the \Index{declaring occurrence} of the identifier.
```
Changed to 'a declaring occurrence'. I think that's the most useful change: We just need to be able to say for any given occurrence of an identifier whether it's a declaring or a referencing occurrence, and then we'll talk about the declaration where it occurs, and the variable that it names, etc.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
+\LMHash{}%
+A \IndexCustom{class variable}{variable!class}
+is a variable whose declaration is
+immediately nested inside a class declaration and
```
Done.

I'd love to call them "static variables", but that phrase is already taken, and it means any non-local variable which is not an instance variable. I think we should just get rid of "static variable" as defined today and start using "static variable" to mean a variable that has the modifier \STATIC.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
-%% want to start talking about a grammar before inference and another one
-%% after inference.
-In these cases, the static type of $v$ is \DYNAMIC{}
+In these cases, the declared type of \id{} is \DYNAMIC{}
```
Type inference is in general a TODO. We talk about type inference approximately 20 times, most of them saying that type inference is assumed to have taken place already. So if a type can be inferred then the declaration doesn't have any of these forms.

I do recognize that we should talk about the program as given (completely unprocessed source), and then we should specify how type inference proceeds, and then we will probably imply that type inference transforms the program, and then we'll describe how the fully type-inferred program runs.

However, that's a large restructuring, so I think we shouldn't try to fix it as part of this PR.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
 (\ref{typeDynamic}).
 
+\commentary{%
+Type inference is assumed to have taken place already
```
Extended the commentary slightly to say that if a type is inferred then the declared type is the inferred type and not `dynamic`.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
+whose execution sets the value of \id{} to the incoming argument $x$.
+
+\commentary{%
+It is again assumed that type inference has taken place already,
```
Again, we should not block this PR on fixing the specification of type inference.

Considering type inference to be a process that annotates syntax with extra data is a very interesting take on it. We would certainly have to switch to indirect denotation of such properties (then 'the declared type' would mean 'the semantic concept of a declared type which could be provided syntactically and could be extra data').

I certainly do not think we should pretend that `\VAR` does not exist.

eernstg replied:
Changed the references to type inference such that it allows for type inference to have occurred already in most places, but also allows for type inference to be a future step when considering variable declarations whose declared type is obtained by type inference.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
-whose execution sets the value of $v$ to the incoming argument $x$.
+\code{\STATIC?\,\,\LATE?\,\,\VAR\,\,\id{} = $e$;}
+implicitly induces a setter with the header
+\code{\VOID\,\,\SET\,\,\id(\DYNAMIC\,\,$x$)},
```
I kept the style from the previous case and added commentary saying that type inference could have provided a type different from `dynamic`.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
+so long as the variable is not known to be definitely assigned.
+The property that the variable is never mutated after initialization
+is enforced dynamically rather than statically.%
+}
```
Removed the commentary.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
+A library variable which introduces a setter will introduce
+a library setter into the enclosing library scope.
+A class variable which introduces a setter will introduce 
+a static setter into the body scope of the immediately enclosing class.
```
It would certainly be nice if we could avoid repeating this list many times, but every new defined word is also a burden on the reader who may need to look up that particular word (and just find a definition that takes 6-8 words, in which case it's faster to read those 4 words than finding that 8 word definition).

So I'm keeping the long list for now. ;-)

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
-the immediately enclosing class.
-A mutable instance variable introduces an instance setter into
-the immediately enclosing class.
+A non-local variable introduces a setter if{}f it is mutable or late and final.
```
Done.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1446,34 +1584,56 @@ \section{Variables}
+or lexical lookup yields nothing, and the location has access to \THIS,
+and the interface of the enclosing class or mixin has
+a setter named \code{\id=},
+it is a compile-time error to assign to \id.
```
The point is that a non-late static immutable variable declaration D named `id` does not introduce a setter, but there could still be a bunch of reasons why `id = e` is not an error at a location where `id` resolves to D. Clarified, simplified, and shortened the text.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1446,34 +1584,56 @@ \section{Variables}
+Similarly, assignment to a non-late final instance variable \id{}
 is a compile-time error,
-unless a setter named \code{$v$=} is in scope,
+unless a setter named \code{\id=} can be found as above,
```
This was rewritten along with the changes mentioned in the previous comment.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1446,34 +1584,56 @@ \section{Variables}
+it is a compile-time error to assign to \id.
 
-Similarly, assignment to a final instance variable $v$
+Similarly, assignment to a non-late final instance variable \id{}
```
Rewritten as mentioned in the comment about line 1596.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1446,34 +1584,56 @@ \section{Variables}
+and in particular it makes no sense to say
+that it has a specific initial value.
+For example, it is an error if a final instance variable
+is not initialized by one of the above mechanisms.%
```
Deleted most of this commentary.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1514,65 +1675,205 @@ \section{Variables}
+\LMHash{}%
+A
+\IndexCustom{late-initialized getter}{getter!late-initialized}
+is a getter $g$ which is implicitly induced by a non-local variable $v$
```
But we are already specializing exactly which declarations have a late-initialized getter elsewhere, saying it here might be precise, but it is also redundant.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1342,240 @@ \section{Variables}
+the library scope of the enclosing library.
 A class variable introduces a static getter into
-the immediately enclosing class.
+the body scope of the immediately enclosing class.
```
Done.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1514,65 +1675,205 @@ \section{Variables}
+An invocation of $g$ proceeds as follows:
+If the variable $v$ has not been bound to an object then
+$e$ is evaluated to an object $o$,
+$v$ is bound to $o$,
```
Good catch! Got it.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -17377,15 +17745,39 @@ \subsection{Local Variable Declaration}
+% The following error can occur due to implicit casts.
 A dynamic type error occurs
-if the dynamic type of $o$ is not a subtype of the actual type
+if the dynamic type of $o$ is not a subtype of the actual declared type
```
Added commentary to say that such transformations are assumed to have taken place already.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1514,65 +1675,205 @@ \section{Variables}
+Any situation where \id{} is  unbound when this getter is evaluated
+is a compile-time error,
+so there is no need to specify a behavior for that situation.%
+}
```
Much better! Done.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -17377,15 +17745,39 @@ \subsection{Local Variable Declaration}
+\begin{dartCode}
+\VOID\ main() \{
+  int i = 0;
+  \LATE\ \FINAL\ int x = i++ == 0 ? x + 1 : 0;
```
Good catch! The conclusion would be that this example goes away, because the issue isn't relevant for a local variable. The corresponding situation is already described for non-local variables in section 8.

eernstg commented:
The approach to type inference is tricky, and we are not going to resolve all issues with that topic in this PR. So I mainly changed the wording at the locations where "type inference is assumed to have taken place already" is obviously wrong. We should deal with that topic by actually specifying type inference in detail, but that won't be right now.

Next, lots of imprecise details have been fixed.

PTAL.

eernstg commented:
Ah, two more updates needed; will be finished this with this hour.

eernstg commented:
OK, latest updates have now been uploaded, PTAL.

eernstg commented:
@lrhn, friendly ping..

eernstg commented:
@lrhn, friendly ping...

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1350,256 @@ \section{Variables}
-induces an implicit setter function with signature
-\code{\VOID{} \SET{} $v$=(\DYNAMIC{} $x$)}
-whose execution sets the value of $v$ to the incoming argument $x$.
+\code{\STATIC?\,\,\LATE?\,\,\VAR\,\,\id;}
```
Should we include a non-local `static? late final id;`?

(Since it really is mutable.)

Otherwise I can see it's in line 1556.

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1350,256 @@ \section{Variables}
 A class variable introduces a static getter into
-the immediately enclosing class.
+%% TODO(eernst): Come enhanced enums, add enum to this list.
+the body scope of the immediately enclosing class, mixin, or extension.
```
This has come to pass, so is this in anticipation of including that in the spec?

Maybe just say `or enum` today. It's not wrong when you can't declare them, and then we won't forget when we do allow it.

Should it be "body scope of .... extension declaration."
(Do we even have a notion of a "class" concept separate from the declaration? Formally?)

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -1514,65 +1692,210 @@ \section{Variables}
+and the evaluation of $g$ completes returning $o$.
+If the evaluation of $e$ throws then
+the invocation of $g$ completes throwing the same object and stack trace,
+and no binding of $v$ to any object occurs.
```
Change to:
```
and does not change the binding of $v$.
```
It's possible for the evaluation of the initializer expression to bind `v` and then throw. It's too late to say that no binding occurs. No *further* binding, possibly, but i like the phrase above better.

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -17363,41 +17707,107 @@ \subsection{Local Variable Declaration}
+
+%% TODO(eernst): Revise when flow analysis is added.
+\commentary{%
+If a local variable $v$ is \FINAL{} and not \LATE,
```
Should this use "is immutable" instead of "is `final`"?

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -1317,127 +1350,256 @@ \section{Variables}
+after \id{} has been initialized.%
+}
+A variable is immutable if{}f its declaration includes
 the modifier \FINAL{} or the modifier \CONST.
```
This is again based on the perspective that a variable is mutable if it induces a setter, even in the case where that setter will ensure at run time that the variable cannot be mutated.

I'd prefer that we only use 'mutable' / 'immutable' in connection with actual mutation, not just with the formal property of having a setter or not.

eernstg commented:
One more thing: As I mentioned, we currently use 'statically allocated variable' rather than the old 'static variable' meaning 'library variable or \STATIC{} variable' (that was 100% likely to confuse anyone who doesn't know). We need a better phrase, but it's very easy to change all occurrences because 'statically allocated variable' isn't used otherwise.

eernstg commented:
Changed the text to avoid using the word 'mutable' and the word 'immutable' anywhere. It is now spelled out in each case which modifiers are required/assumed.

Changed the text to avoid using 'statically allocated variable' anywhere. Now spells out every time that the variable is a library variable or a static variable (and that's the new meaning of 'static variable', namely: A variable whose declaration includes the word `static`). In several locations we now use 'non-local variable' where we previously had all kinds of non-local variables anyway (as in 'if $v$ is a library variable, a static variable (new meaning), or an instance variable' becomes 'if $v$ is a non-local variable').

Eliminated the phrase 'static function' (there was exactly one, in non-normative text). Now using 'static member', which is well defined (meaning a member declaration that includes the keyword `static`).
